axlearn/common/file_system.py

```diff
@@ -173,7 +208,59 @@ def makedirs(path: str):
     return tf.io.gfile.makedirs(path)
 
 
+@functools.cache
+def _is_hierarchical_namespace_enabled(bucket_name: str) -> bool:
+    """Return whether hierarchical namespace is enabled."""
+    client = _gs_client()
+    bucket = client.get_bucket(bucket_name)
+    return bucket.hierarchical_namespace_enabled
+
+
+def _rm_empty_folders(bucket: str, prefix: str):
+    """For a hierarchical namespace bucket, delete empty folders recursively."""
+    # pylint: disable=import-outside-toplevel
+    from google.cloud import storage_control_v2
+
+    client = _gs_control_client()
+    project_path = client.common_project_path("_")
+    bucket_path = f"{project_path}/buckets/{bucket}"
+    folders = set(
+        folder.name  # Format: "projects/{project}/buckets/{bucket}/folders/{folder}"
+        for folder in client.list_folders(
+            request=storage_control_v2.ListFoldersRequest(
+                parent=bucket_path, prefix=prefix.strip("/") + "/"
+            )
+        )
+    )
+
+    with concurrent.futures.ThreadPoolExecutor() as pool:
+        # Delete empty folders first, otherwise GCS will complain:
+        # FailedPrecondition: 400 The folder you tried to delete is not empty.
+        while len(folders) > 0:
+            parents = set(os.path.dirname(x.rstrip("/")) + "/" for x in folders)
+            leaves = folders - parents
+            res = list(
+                pool.map(
+                    client.delete_folder,
+                    [storage_control_v2.DeleteFolderRequest(name=f) for f in leaves],
+                )
+            )
+            folders = folders - leaves
+            logging.info(
+                "Deleted %s folders, %s remaining. [%s][%s]", len(res), len(folders), bucket, prefix
+            )
+
+
 @_wrap_tf_errors
 def rmtree(path: str):
-    """Analogous to tf.io.gfile.rmtree."""
-    return tf.io.gfile.rmtree(path)
+    """Analogous to tf.io.gfile.rmtree.
+    For a hierarchical namespace bucket, all the folders will also be removed."""
+    tf.io.gfile.rmtree(path)
```

Comment:
I'm confused---gfile.rmtree is not enough for HNS? Why do we need to call `_rm_empty_folders`? Can you add a comment?

Reply:
> `gfile.rmtree` is not enough for HNS?

No, tfds doesn't recognize HNS at the moment. It's also reported to the GCS team.

> Can you add a comment?

Updated.